Add write delay strategy && Optimize/Fix queue

vitorenesduarte opened the pull request.

fermadeiral commented:
Hi @vitorenesduarte,

I'm collecting bugs for research. The tool I use for that had identified your commit [1] as fixing `NullPointerException` with the patch applied in `CommittedQueueBox.java`. However, the message of such commit is "Refactor tests". I would like to ask you if there exists a bug fix too in that commit I mentioned.

Thank you in advance!

[1] https://github.com/vitorenesduarte/VCD-java-client/pull/27/commits/bba1c44e155e913e7108fa4c218915edd906e65a

vitorenesduarte commented:
Hi @fermadeiral 

I'm (or was) in the middle of a bug-hunt.
The patch on `CommittedQueueBox.java` indeed fixes a `NullPointerException` which resulted from "refactoring" the tests. I noticed I was artificially creating an argument that should be `null` for these tests. Once I changed that (the image shows a subset of the whole patch in that commit showing this change)
<img width="995" alt="screen shot 2018-10-05 at 13 25 19" src="https://user-images.githubusercontent.com/4520516/46535159-6afca700-c8a2-11e8-95ce-4bc70748ff60.png">
I got the `NullPointerException` and fixed it.

Now, if you want bugs, I just fixed two in [2] and [3].
- [2] appeared in commit [1]. Commit [1] optimizes Tarjan's algorithm by only checking the highest dependencies (if a node depends on 5, it depends on 4, and so on, and thus, it's enough to check the highest one) of each node in the graph (that I'm calling frontier).
<img width="950" alt="screen shot 2018-10-05 at 13 32 30" src="https://user-images.githubusercontent.com/4520516/46535444-4e14a380-c8a3-11e8-89f7-216d2982bce1.png">
However, the implementation of `frontier` had a naive bug, which you see fixed in [2]

- [3] has been here for a long time, but only now I realized it's incorrect. Sometimes this algorithm would get stuck, I was attributing it to something else in the stack, but this might be the reason.
After finding an SCC with Tarjan's alg., we may be able to find another one. For that, we need to check the next sequence. The implementation of `next` was wrong in class `ExceptionSet` (that represents a set of integers as the highest one, and a set of exceptions). This is the fix:
<img width="988" alt="screen shot 2018-10-05 at 13 31 20" src="https://user-images.githubusercontent.com/4520516/46535630-fa568a00-c8a3-11e8-8277-84639a1c67a9.png">

Let me know if you need more details.
And, good luck with that paper!

[1] 6fb2081e7bd5dab2569b51a6348bc52222871fc9
[2] 9c15bd8edc417e8122d169691a23be2d0945fcef
[3] ed4467e27433ed53a80a7c5aa27c0aacd3f1eda5

fermadeiral commented:
Hi @vitorenesduarte.

> I'm (or was) in the middle of a bug-hunt.
The patch on CommittedQueueBox.java indeed fixes a NullPointerException which resulted from "refactoring" the tests. I noticed I was artificially creating an argument that should be null for these tests. Once I changed that (the image shows a subset of the whole patch in that commit showing this change)

Thank you a lot for your explanation! I'm happy that there was actually a bug there!

> Now, if you want bugs, I just fixed two in [2] and [3]. [...]

I will look these cases carefully.

> Let me know if you need more details.

For now it was just that question, but I really appreciate your kindness.

> And, good luck with that paper!

:-)
